File: events/year_2020/day_04/day_04.py
"""
Advent of Code - Year 2020 - Day 4
https://adventofcode.com/2020/day/4
"""

import string


def preprocessing(puzzle_input: str) -> list[dict[str, str]]:
    """
    Parses the puzzle input string into a list of dictionaries, each representing a passport with
    its fields as key-value pairs.
    """
    details = []
    for passport in puzzle_input.split("\n\n"):
        passport = passport.replace("\n", " ")
        infos = {}
        for info in passport.split(" "):
            key, value = info.split(":")
            infos[key] = value
        details.append(infos)
    return details


def solver(passports: list[dict[str, str]]) -> tuple[int, int]:
    """
    Counts the number of candidate and valid passports from a list of passport dictionaries
    according to specific field and validation rules.
    """
    candidates = 0
    valids = 0

    for passport in passports:
        if len(passport) + ("cid" not in passport.keys()) == 8:
            candidates += 1
            if is_valid(passport):
                valids += 1

    return candidates, valids


def is_valid(passport: dict[str, str]) -> bool:
    """
    Checks if a passport dictionary contains valid field values according to specific validation
    rules for each field.
    """
    return all((
        1920 <= int(passport["byr"]) <= 2002,
        2010 <= int(passport["iyr"]) <= 2020,
        2020 <= int(passport["eyr"]) <= 2030,
        any((
            "in" in passport["hgt"] and 59 <= int(passport["hgt"][:-2]) <= 76,
            "cm" in passport["hgt"] and 150 <= int(passport["hgt"][:-2]) <= 193)),
        passport["hcl"].startswith("#") and all(c in string.hexdigits for c in passport["hcl"][1:]),
        passport["ecl"] in ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"],
        len(passport["pid"]) == 9,))

File: events/year_2020/day_04/day_04.rs
aoc::main!();

fn preprocessing(input: &str) -> Vec<HashMap<&str, &str>> {
    input
        .split("\n\n")
        .map(|passport| {
            passport
                .split(&['\n', ':', ' '])
                .collect_vec()
                .chunks(2)
                .map(|chunk| (chunk[0], chunk[1]))
                .collect::<HashMap<_, _>>()
        })
        .collect_vec()
}

fn part_1(passports: Vec<HashMap<&str, &str>>) -> usize {
    passports
        .iter()
        .filter(|passport| {
            passport.len() == 8 || (passport.len() == 7 && !passport.contains_key("cid"))
        })
        .count()
}

fn part_2(passports: Vec<HashMap<&str, &str>>) -> usize {
    passports
        .iter()
        .filter(|passport| {
            (passport.len() == 8 || (passport.len() == 7 && !passport.contains_key("cid")))
                && (1920..=2002).contains(&passport["byr"].parse::<u16>().unwrap())
                && (2010..=2020).contains(&passport["iyr"].parse::<u16>().unwrap())
                && (2020..=2030).contains(&passport["eyr"].parse::<u16>().unwrap())
                && ((passport["hgt"].contains("in")
                    && (59..=76).contains(&passport["hgt"][..2].parse::<u16>().unwrap_or(0)))
                    || (passport["hgt"].contains("cm")
                        && (150..=193).contains(&passport["hgt"][..3].parse::<u16>().unwrap_or(0))))
                && (passport["hcl"].starts_with('#')
                    && passport["hcl"][1..].chars().all(|c| c.is_ascii_hexdigit()))
                && ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"].contains(&passport["ecl"])
                && passport["pid"].len() == 9
        })
        .count()
}
